grid-compose/internal/app.go

```diff
@@ -0,0 +1,372 @@
+			}
+
+			dl.Vms = append(dl.Vms, vm)
+			dl.Disks = append(dl.Disks, disks...)
+		}
+
+		dls = append(dls, dl)
+	}
+
+	log.Info().Str("status", "started").Msg("deploying networks...")
+
+	for _, network := range networks {
+		if err := a.Client.NetworkDeployer.Deploy(ctx, network); err != nil {
+			return err
+		}
+	}
+	log.Info().Str("status", "done").Msg("networks deployed successfully")
+
+	deployed := make([]*workloads.Deployment, 0)
+
+	for _, dl := range dls {
+		log.Info().Str("deployment", dl.Name).Msg("deploying...")
+
+		if err := a.Client.DeploymentDeployer.Deploy(ctx, dl); err != nil {
+			log.Info().Msg("an error occurred while deploying the deployment, canceling all deployments")
+
+			for _, network := range networks {
+				if err := a.Client.NetworkDeployer.Cancel(ctx, network); err != nil {
+					return err
+				}
+			}
+
+			for _, deployment := range deployed {
+				if err := a.Client.DeploymentDeployer.Cancel(ctx, deployment); err != nil {
+					return err
+				}
+			}
+			log.Info().Msg("all deployments canceled successfully")
+			return err
+		}
+		log.Info().Str("deployment", dl.Name).Msg("deployed successfully")
+
+		deployed = append(deployed, dl)
+	}
+
+	log.Info().Msg("all services deployed successfully")
+	return nil
+}
+
+// Ps lists the deployed services
+func (a *App) Ps(ctx context.Context, flags *pflag.FlagSet) error {
+	verbose, outputFile, err := parsePsFlags(flags)
+	if err != nil {
+		return err
+	}
+
+	var output strings.Builder
+	outputMap := make(map[string]gridtypes.Deployment)
+
+	if !verbose {
+		output.WriteString(fmt.Sprintf("%-15s | %-15s | %-15s | %-15s | %-10s | %s\n", "Deployment Name", "Network", "Services", "Storage", "State", "IP Address"))
+		output.WriteString(strings.Repeat("-", 100) + "\n")
+	}
+
+	for networkName, deploymentData := range a.Config.DeploymentData {
+		projectName := a.getProjectName(networkName)
+
+		if err := a.loadCurrentNodeDeployments(projectName); err != nil {
+			return err
+		}
+
+		contracts, err := a.Client.ContractsGetter.ListContractsOfProjectName(projectName)
+		if err != nil {
+			return err
+		}
+
+		for _, contract := range contracts.NodeContracts {
+			contractDlData, err := workloads.ParseDeploymentData(contract.DeploymentData)
+			if err != nil {
+				return err
+			}
+
+			if contractDlData.Type == "network" {
+				continue
+			}
+
+			dlAdded := false
+			for _, service := range deploymentData.Services {
+				wl, dl, err := a.Client.State.GetWorkloadInDeployment(ctx, contract.NodeID, service.Name, contractDlData.Name)
+				if err != nil {
+					return err
+				}
+
+				vm, err := workloads.NewVMFromWorkload(&wl, &dl)
+				if err != nil {
+					return err
+				}
+
+				if !verbose {
+					if !dlAdded {
+						output.WriteString(fmt.Sprintf("%-15s | %-15s | %-15s | %-15s | %-10s | %s\n", contractDlData.Name, vm.NetworkName, vm.Name, vm.Mounts[0].DiskName, wl.Result.State, utils.GetVmAddresses(vm)))
```

Comment:
could this formatting be abstracted? get the result as a map/array then pass it to some method maybe in `pkg/logging/print_resutl.go` it will be easier to make changes or even use a different pkg for formatting